scanpy/plotting/_anndata.py

```diff
@@ -1865,7 +1865,15 @@ def _prepare_dataframe(
         # join the groupby values  using "_" to make a new 'category'
         categorical = obs_tidy[groupby].agg('_'.join, axis=1).astype('category')
         categorical.name = "_".join(groupby)
-
+        # preserve category order
+        categories = (
+            obs_tidy[groupby]
+            .drop_duplicates()
+            .sort_values(groupby)
+            .agg('_'.join, axis=1)
+            .values
+        )
+        categorical.cat.reorder_categories(categories, inplace=True)
```

Comment:
Might be nice to figure out categories in a way that scales with number of categories, not number of samples. Maybe something like:


```python
from itertools import product

order = {
    "_".join(k): idx for idx, k in 
    enumerate(product(*(obs_tidy[g].cat.categories for g in groupby)))
}
categorical = categorical.cat.reorder_categories(
    sorted(categorical.cat.categories, key=lambda x: order[x])
)
```

I do wish "categorical" weren't so many characters. Makes these lines so long!